lib/Dialect/SSP/SSPOps.cpp

```diff
@@ -49,6 +49,255 @@ DependenceGraphOp InstanceOp::getDependenceGraph() {
   return *getOps<DependenceGraphOp>().begin();
 }
 
+//===----------------------------------------------------------------------===//
+// OperationOp
+//===----------------------------------------------------------------------===//
+
+ParseResult OperationOp::parse(OpAsmParser &parser, OperationState &result) {
+  auto &builder = parser.getBuilder();
+
+  // Special handling for the linked operator type property
+  SmallVector<Attribute> alreadyParsed;
+
+  if (parser.parseLess())
+    return failure();
+
+  FlatSymbolRefAttr oprRef;
+  auto parseSymbolResult = parser.parseOptionalAttribute(oprRef);
+  if (parseSymbolResult.hasValue()) {
+    assert(succeeded(*parseSymbolResult));
+    alreadyParsed.push_back(builder.getAttr<LinkedOperatorTypeAttr>(oprRef));
+  }
+
+  if (parser.parseGreater())
+    return failure();
+
+  // (Scheduling) operation's name
+  StringAttr opName;
+  (void)parser.parseOptionalSymbolName(opName, SymbolTable::getSymbolAttrName(),
+                                       result.attributes);
+
+  // Dependences
+  SmallVector<OpAsmParser::UnresolvedOperand> unresolvedOperands;
+  SmallVector<Attribute> dependences;
+  unsigned operandIdx = 0;
+  auto parseDependenceSourceWithAttrDict = [&]() -> ParseResult {
+    llvm::SMLoc loc = parser.getCurrentLocation();
+    FlatSymbolRefAttr sourceRef;
+    ArrayAttr properties;
+
+    // Try to parse either symbol reference...
+    auto parseSymbolResult = parser.parseOptionalAttribute(sourceRef);
+    if (parseSymbolResult.hasValue())
+      assert(succeeded(*parseSymbolResult));
+    else {
+      // ...or an SSA operand.
+      OpAsmParser::UnresolvedOperand operand;
+      if (parser.parseOperand(operand))
+        return parser.emitError(loc, "expected SSA value or symbol reference");
+
+      unresolvedOperands.push_back(operand);
+    }
+
+    // Parse the properties, if present.
+    parseOptionalPropertyArray(properties, parser);
+
+    // No need to explicitly store SSA deps without properties.
+    if (sourceRef || properties)
+      dependences.push_back(
+          builder.getAttr<DependenceAttr>(operandIdx, sourceRef, properties));
+
+    ++operandIdx;
+    return success();
+  };
+
+  if (parser.parseCommaSeparatedList(AsmParser::Delimiter::Paren,
+                                     parseDependenceSourceWithAttrDict))
+    return failure();
+
+  if (!dependences.empty())
+    result.addAttribute(builder.getStringAttr("dependences"),
+                        builder.getArrayAttr(dependences));
+
+  // Properties
+  ArrayAttr properties;
+  auto parsePropertiesResult =
+      parseOptionalPropertyArray(properties, parser, alreadyParsed);
+  if (parsePropertiesResult.hasValue()) {
+    if (failed(*parsePropertiesResult))
+      return failure();
+    result.addAttribute(builder.getStringAttr("properties"), properties);
+  }
+
+  // Parse default attr-dict
+  (void)parser.parseOptionalAttrDict(result.attributes);
```

Comment:
Is this the correct way of handling `parseOptionalAttrDict`? It doesn't return a `OptionalParseResult` so I'm assuming if `parseOptionalAttrDict` returns failure, something actually failed.

Reply:
You're right, thanks!